src/sentry/api/endpoints/project_codeowners.py

```diff
@@ -1,11 +1,156 @@
+import logging
+
+from django.db import IntegrityError
+from django.utils import timezone
+
+from rest_framework import serializers, status
 from rest_framework.response import Response
 
 from sentry.api.bases.project import ProjectEndpoint
 from sentry.api.serializers import serialize
-from sentry.models import ProjectCodeOwners
+from sentry.models import (
+    ProjectCodeOwners,
+    ExternalUser,
+    ExternalTeam,
+    RepositoryProjectPathConfig,
+    UserEmail,
+)
+from sentry.api.serializers.rest_framework.base import CamelSnakeModelSerializer
+from sentry.ownership.grammar import (
+    parse_code_owners,
+    convert_codeowners_syntax,
+)
+
+from sentry.api.endpoints.project_ownership import ProjectOwnershipSerializer, ProjectOwnershipMixin
+
+logger = logging.getLogger(__name__)
+
+
+class ProjectCodeOwnerSerializer(CamelSnakeModelSerializer):
+    code_mapping_id = serializers.IntegerField(required=True)
+    raw = serializers.CharField(required=True)
+    organization_integration_id = serializers.IntegerField(required=False)
+
+    class Meta:
+        model = ProjectCodeOwners
+        fields = ["raw", "code_mapping_id", "organization_integration_id"]
+
+    def validate(self, attrs):
+        # If it already exists, set default attrs with existing values
+        if self.instance:
+            attrs = {
+                "raw": self.instance.raw,
+                "code_mapping_id": self.instance.repository_project_path_config,
+                **attrs,
+            }
+
+        if not attrs.get("raw", "").strip():
+            return attrs
+        external_association_err = []
+        # Get list of team/user names from CODEOWNERS file
+        teamnames, usernames, emails = parse_code_owners(attrs["raw"])
+
+        # Check if there exists Sentry users with the emails listed in CODEOWNERS
+        user_emails = UserEmail.objects.filter(email__in=emails)
+        user_emails = [user.email for user in user_emails]
+        user_emails_diff = [email for email in emails if email not in user_emails]
+
+        if len(user_emails_diff):
```

Comment:
I wonder if we could pull out some of the logic into like `_validate_association()`

```python
def _validate_association(raw_items, associations, type):
    if type == "emails":
         sentry_items = [item.email for item in associations]
    else: 
         sentry_items = [item.external_name for item in associations]

    diff = [item for item in raw_items if not item not in sentry_items]

    if len(diff):
        return [f'The following %s do not have an association in Sentry: {", ".join(diff)}]
    
    return []
```
    
where the raw items would be the values from the `parse_code_owners` the types would be `emails`, `usernames` `teamnames` and associations would be the sentry querysets

we don't have to do something like this but it does feel like there is a lot of repeated code that is very similar. Also I would say we could raise early, but im guessing that we rather give them all the missing associations instead of just the first set so that they can fix them all at once

Reply:
@MeredithAnya Good point on the duplicated filter logic! Yeah I don't think it'll be a good DX to raise early. I think its preferable to return all the missing associations at once and let them fix at once.